R/file_utils.R

```diff
@@ -574,17 +586,66 @@ load_twas_weights <- function(weight_db_files, conditions = NULL,
                               variable_name_obj = c("preset_variants_result", "variant_names"),
                               susie_obj = c("preset_variants_result", "susie_result_trimmed"),
                               twas_weights_table = "twas_weights") {
+
   ## Internal function to load and validate data from RDS files
-  load_and_validate_data <- function(weight_db_files, conditions, variable_name_obj) {
-    all_data <- lapply(weight_db_files, readRDS)
-    unique_regions <- unique(unlist(lapply(all_data, function(data) names(data))))
-    # Check if region from all RDS files are the same
+  load_and_validate_data <- function(weight_db_files, conditions, variable_name_obj){
+    all_data <- do.call(c, lapply(unname(weight_db_files), function(rds_file) {
+      db <- readRDS(rds_file)
+      if ("mnm_result" %in% names(db)) {
+        db <- list(mnm_rs=db)
+      } else {
+        gene <- names(db)
+        # Check if region from all RDS files are the same
+        if (length(gene) != 1) {
+          stop("More than one region provided in the RDS file. ")
+        } else {
+          names(db[[gene]]) <- sapply(names(db[[gene]]), function(context) clean_context_names(context, gene))
+        }                                            
+      }
+      return(db)    
+    }))
+    #Check if region from all RDS files are the same
+    unique_regions <- unique(names(all_data)[!names(all_data) %in% 'mnm_rs'])
     if (length(unique_regions) != 1) {
       stop("The RDS files do not refer to the same region.")
     } else {
-      # Assuming all data refer to the same region, now combine data by conditions
-      combined_all_data <- do.call("c", lapply(all_data, function(data) data[[1]]))
+      # Combine the lists with the same region name
+      combined_all_data <- lapply(split(all_data, names(all_data)),function(lst) {
+         result <- list()
+          for (item in lst) {
+            for (name in names(item)) {
+              if (name %in% names(result)) {
+                if (result[[name]] != item[[name]]) {
+                  stop("Different twas weight data provided for identical context name. ")
+                }
+              } else {
+                result[[name]] <- item[[name]]
+              }
+            }
+          }
+          return(result)
+        })
+    }                                                                   
+    # merge univariate and multivariate results for same gene-context pair                                                         
+    if ("mnm_rs" %in% names(combined_all_data)){
```

Comment:
This needs some clean-up still. I think this can be more generic without using if ("mnm_rs" %in% names(combined_all_data)) because the task is essentailly to make it work allowing for missing variants. But we can table it.